pe/orchestration.py

```diff
@@ -0,0 +1,246 @@
+            LOGGER.info('No records will be updated in the database')
+            return None
+
+        resp_data: Dict[str, Any] = json.loads(response.text)
+        LOGGER.debug(resp_data)
+
+        schema_name: str = 'putPlcExamScoreResponse'
+        results: Dict[str, Any] = resp_data[schema_name][schema_name]
+
+        if results['BadCount'] > 0:
+            LOGGER.warning(f"Discovered {results['BadCount']} record error(s): {results['Errors']}")
+
+        # Hope this can be simplified in the future if API response data can be made to use consistent types
+        if results['GoodCount'] > 1:
+            success_uniqnames: List[str] = [success_dict['uniqname'] for success_dict in results['Success']]
+        elif results['GoodCount'] == 1:
+            success_uniqnames = [results['Success']['uniqname']]
+        else:
+            success_uniqnames = []
+
+        if len(success_uniqnames) == 0:
+            LOGGER.warning('No scores were transmitted successfully.')
+        else:
+            timestamp: datetime = datetime.now(tz=utc)
+
+            subs_to_update = []
+            for sub in subs_to_send:
+                if sub.student_uniqname in success_uniqnames:
+                    sub.transmitted = True
+                    sub.transmitted_timestamp = timestamp
+                    subs_to_update.append(sub)
+            Submission.objects.bulk_update(objs=subs_to_update, fields=['transmitted', 'transmitted_timestamp'])
+            LOGGER.info(f'Transmitted {len(subs_to_update)} score(s) successfully and updated submission record(s).')
+        return None
+
+    def main(self) -> None:
+        """
+        High-level process method for class. Pulls Canvas data, sends data, and logs activity in the database.
+
+        :return: None
+        :rtype: None
+        """
+        # Fetch data from Canvas API and store as submission records in the database
+        sub_dicts: List[Dict[str, Any]] = self.get_sub_dicts_for_exam()
+        if len(sub_dicts) > 0:
+            self.create_sub_records(sub_dicts)
+
+        # Find old and new submissions for exam to send to M-Pathways
+        sub_to_transmit_qs: QuerySet = self.exam.submissions.filter(transmitted=False)
+        subs_to_transmit: List[Submission] = list(sub_to_transmit_qs.all())
+
+        # Identify old submissions for debugging purposes
+        redo_subs: List[Submission] = list(sub_to_transmit_qs.filter(graded_timestamp__lt=self.sub_time_filter))
+        if len(redo_subs) > 0:
+            LOGGER.info(f'Will try to re-send {len(redo_subs)} previously un-transmitted submissions')
+            if len(redo_subs) > 10:
+                LOGGER.debug(f'First 10 previously un-transmitted submissions: {redo_subs[:10]}')
+            else:
+                LOGGER.debug(f'All previously un-transmitted submissions: {redo_subs}')
+
+        # Identify and separate submissions to send with duplicate uniqnames
```

Comment:
Yes, thanks for asking, @pushyamig, I was going to check in with you about my thinking here. I don't actually know if this is necessary, I keep going back and forth about it.

You can see more or less what I had before in this commit, in the changes to `pe/orchestration`: https://github.com/tl-its-umich-edu/placement-exams/pull/12/commits/74cb3e0ebdd140ada92eb6f80f5d641b94b33e99#

Before, when I was processing the response from M-Pathways, I was just updating all submissions for the exam that were un-transmitted and had a uniqname that came back nested under "Success". This would not be a problem if we assume that there will never be duplicate uniqnames in scores for a single exam that are sent. However, I'm not sure if that's an assumption we can make; if it is, then I can revert back to something like I had before.

Because I don't fully understand in what situations duplicates can occur -- and I thought maybe processes might change in the future -- I guess I wanted to try to "future proof" this to handle duplicates. My concern was that, if we needed to send multiple scores, and for some reason, one of them failed to be sent (because of mis-formatting or whatever), I would have no way of knowing which failed and which succeeded. I would see the same uniqname and placement score under "Success" and "Error"; there is no way of uniquely identifying a submission/score when it comes back from the API. So my solution presented in the PR now is to identify these duplicates beforehand, and then send them one at a time, so that we know exactly which Submission the response refers to. 

Again, the probably that this case would occur is really small, and maybe zero. The student would have to have multiple submissions within a short time frame. On some level, I feel like the application should be able to handle duplicates, but I am totally willing to disregard this and revert back to something simpler.

Reply:
Okay, I think we agreed this approach is okay for now. I will keep it as is, but we can revisit if necessary.